CES scale test: prelim testing for 5k scale test

thorn3r opened the pull request:
For testing
Swaps out the create-cluster action with a [modified copy on my fork](https://github.com/cilium/scale-tests-action/pull/7) to adjust KCM and scheduler settings to achieve higher scale.

thorn3r committed:
CES scale test: prelim testing for 5k scale test

Modify 100 nodes scale test to prepare for testing a 5000 nodes cluster.
Updates:
- Use forked create-cluster action that adjusts KCM and scheduler QPS
- Lower root disk size of nodes to 50GB, down from 64GB/128GB for
  control-plane/node
- Disable kube-proxy and set the K8s Service host and port to the
  address of the apiserver loadbalancer.
- Raise Prometheus scrape interval to 60s (default 30s).
- Upload sysdump on failure

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -160,14 +161,16 @@ jobs:
+          control_plane_volume_size: 50
           node_size: e2-medium
-          node_count: 100
+          node_count: 20
```
FYI You were trying to run 20 nodes with a test targeted for 100 nodes (`--nodes=100 \` below) so there was not enough room for pods and test timed out.

thorn3r replied:
I only wanted to test out the changes I made to reduce the root disk size without wasting so many nodes. Each node was being deployed with 128GB disk which was eating up our quota very quickly (and costing more $)

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -217,6 +220,8 @@ jobs:
           export CL2_PROMETHEUS_SCRAPE_CILIUM_OPERATOR=true
           export CL2_PROMETHEUS_SCRAPE_CILIUM_AGENT=true
           export CL2_PROMETHEUS_MEMORY_SCALE_FACTOR=2.0
+          export CL2_LOAD_TEST_THROUGHPUT=50.0
```
Commenting on random line:

Currently, we are running kube-proxy in the cluster. For 5k I think it makes sense to stick with KPR and disabling kube-proxy - services will generate quite a lot of churn. 

Even for 100 nodes it still makes sense to disable it so it can be a totally separate PR.

thorn3r replied:
Good call :+1:

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -88,7 +88,8 @@ jobs:
             --helm-set=hubble.relay.image.tag=${SHA} \
-            --helm-set=hubble.relay.image.useDigest=false"
+            --helm-set=hubble.relay.image.useDigest=false \
+            --helm-set=ciliumEndpointSlice.enabled=true"
```
I'm pretty sure we will kill prometheus if we don't disable `endpointHealthChecking.enabled=false` and `healthChecking=false` - just like we recommend in scalability report now. I will create an issue for it to fix it once and for all.  Link: TBD.

marseel replied:
FYI, I've started issue with some proposal: https://github.com/cilium/cilium/issues/32820

thorn3r replied:
excellent proposal :+1:

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -160,14 +161,16 @@ jobs:
           node_size: e2-medium
-          node_count: 100
+          node_count: 20
+          node_volume_size: 50
```
`node_size: e2-standard-8` for heavy-deployment will need to be `e2-standard-32` probably:
https://github.com/kubernetes/test-infra/blob/8c94170825a87c6b6eb339a117263f6a214d0872/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L99C37-L99C51

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -160,14 +161,16 @@ jobs:
           cluster_name: ${{ steps.vars.outputs.cluster_name }}
-          control_plane_size: n1-standard-8
-          control_plane_count: 1
+          control_plane_size: n2-standard-64
```
I just double-checked, and upstream runs  `n2-standard-32`. 
From recent test:
```
gce-scale-cluster-master  us-east1-b  n2-standard-32               10.40.0.3    35.237.108.192  RUNNING
```
https://github.com/kubernetes/test-infra/blob/a4f43f0f14feb4ec582db4c8b4c64ac6ca767b5d/config/jobs/kubernetes/sig-scalability/sig-scalability-presets.yaml#L153

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -217,6 +220,8 @@ jobs:
           export CL2_PROMETHEUS_SCRAPE_CILIUM_AGENT=true
           export CL2_PROMETHEUS_MEMORY_SCALE_FACTOR=2.0
+          export CL2_LOAD_TEST_THROUGHPUT=50.0
+          export CL2_DELETE_TEST_THROUGHPUT=50.0
```
FYI, depending on how much metrics we fetch from cilium-agent, we might want to tune the scrape interval:
https://github.com/kubernetes/perf-tests/pull/2651/files#diff-6a4d7ca416707ad8bf902645ddddde590e8fde0c2711e67ee7e68ea38e593d61R3

I've set it to 30s by default, but I am pretty sure it won't be enough.

thorn3r committed:
l2-discovery: fix health reporting for link updater

As-is, when l2 neighbor discovery is enabled, the
node-neighbor-link-updater controller fails with "invalid node spec
found in queue". This is due to a bug in the controller's DoFunc, where
an empty list is treated the same as an invalid queue entry.

When this controller fails, `cilium status` reports errors for all nodes
in the cluster similar to the following:

```
cilium             cilium-mgstt    controller node-neighbor-link-updater is failing since 21s (49x): invalid node spec found in queue: (*manager.nodeQueueEntry)(nil)
```

To differentiate between an empty queue and a nil item, the queue's
`pop` method now also returns a bool to indicate whether an element was
successfully retrieved from the queue.

Fixes: 8d525fe39113 ("Add node neighbor link updater controller")

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scaltetest: add step to gather debug info on failure

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
bump prometheus memory constraints

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
always run cluster cleanup

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
cl2-modules: move metrics cardinality to its own action

The CiliumMetricsCardinality measurement consumes a lot of memory. Move
this measurement to its own module to allow including it (or more likely
skipping it) separately from other measurements.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
CES/helm: configure --ces-slice-mode via Helm values

Add ciliumEndpointSlice.slicingMode to support configuring
--ces-slice-mode via Helm values.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scaletest: use FCFS slice mode

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scaletest: 500 nodes

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: only get sysdump on failure

Capturing a sysdump on a large cluster is slow and expensive. Let's
refrain from doing so unless a failure is encountered.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: add measurements for api requests and CEP propagation delay

Add measurements for kube-apiserver requests from Cilium and
CiliumEndpointSlice/CiliumEndpoint propagation delay.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scaletest: 1000 nodes

- Increase node count to 1000
- Increase timeouts on the job, cluster creation, cluster ready, and CL2
  steps.
- Bump create-cluster action to latest version to disable KCM node CIDR
  allocation.
- Increase cilium-operator k8s client QPS/Burst to 100/200
- Increase CiliumEndpointSlice controller's workqueue QPS/Burst to 100/200

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: add count of CiliumEndpoints and CiliumEndpointSlices

Adds 2 new measurements to track the total number of CiliumEndpoints and
CiliumEndpointSlices during during the scale-test.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scaletest: use larger runner

At 1000 nodes the runner is running out of disk space. Standard runners
only have 20GB of disk, which will be increased to 150GB now using the
larger runner type.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: 2000 nodes

Increase node count to 2000 for scale-test

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: use multiple instance-groups

GCE supports a maximum of 1000 nodes in a managed instance group. Due to
this limitation, we need to use additional instance-groups for clusters
larger than 1000 nodes.

The forked create-instance-group action should be reverted once the
changes are merged into main. This was done to decrease the size of the
root volumes.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase validate-cluster timeout to 90m

The validate-cluster action is failing at the hardcoded 20m timeout.
This changes the action to a forked version to allow passing in a
configurable timeout.

related: https://github.com/cilium/scale-tests-action/pull/12

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: 3000 nodes

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: use 8core-32gb runner

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: don't gather debug info

This should be ported to the finaly 5k nodes workflow. Gathering debug
info and sysdump on such a large cluster takes a long time and is
resource intensive. We should avoid this.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase master size to n2-standard-64

See lots of errors near the end of CL2 while cleaning up resources:

```
2024-07-11T18:23:36.3107990Z namespace test-q2kq8l-22 object medium-statefulset-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3109033Z namespace test-q2kq8l-1 object big-job-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3110015Z namespace test-q2kq8l-10 object big-job-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3111046Z namespace test-q2kq8l-7 object small-statefulset-0 updating error: etcdserver: mvcc: database space exceeded
```

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase etcd volume size to 50gb

Bump create-cluster action to include a change to increase etcd volume
size to 50gb. In a 3k nodes test the database was full:

```
2024-07-11T18:23:36.3107990Z namespace test-q2kq8l-22 object medium-statefulset-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3109033Z namespace test-q2kq8l-1 object big-job-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3110015Z namespace test-q2kq8l-10 object big-job-0 updating error: etcdserver: mvcc: database space exceeded
2024-07-11T18:23:36.3111046Z namespace test-q2kq8l-7 object small-statefulset-0 updating error: etcdserver: mvcc: database space exceeded
```

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: re-disable kube-proxy

kube-proxy was inadvertently enabled after bumping cluster-create
version.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
testing: remove anti-affinity rules for cilium-agent

scale-test is failing at 5000 nodes with timeout due to some pods
pending or terminating. Removing the pod anti-affinity rules for the
cilium-agent daemonset, as K8s docs mention this puts a burden on
scheduler performance in large clusters.

Related: https://github.com/cilium/cilium/issues/33462

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: disable KCM node CIDR allocation

This change was unintentionally dropped once the upstream PR was merged.
The default is true, disabling node cidr allocation now needs to be
specified when using the action.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase etcd backend quota

Bumps create-cluster to latest on fork to set etcd backend quota to 8gb.
This should avoid filling up the etcd database, as seen in runs at 3k
nodes.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
disable cilium debug logging

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: separate deletion into multiple steps

When deleting a cluster, kops issues a large volume of API list and read
requests in a short period of time. This is causing us to continually
hit GCE API quota limits, and the problem will compound as we grow the
cluster larger.

This separates the cluster deletion step into multiple phases. Deletion
of each instance group is handled individually, similar to creation.
This should curb some of kops's API usage and spread it over a larger
amount of time.

Additionally, the `delete-instance-group` implements a retry with a 2min
backoff, ensuring that retry after giving the quota time to recover if
the limit is exceeded.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: replace scale-tests-action fork with upstream

Replace the forked scale test actions with the upstream repo at main.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: 5000 nodes

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase master volumes, decrease node volumes

Upstream tests use 1000gb for masters and 30gb for nodes.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: drop step to wait for Cilium to be ready

Running `cilium status` on a large cluster takes a very long time to
return (~1.5 hours at 4k nodes). Since we are already validating that
all Nodes are ready via validate-cluster, it should be safe to assume
that all cilium-agent Pods are also ready.

Let's skip this step to avoid making the workflow longer than it needs
to be.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
scale-test: increase Job and CL2 timeouts

Previous run cut off CL2 at 3 hours due to exceeding the Job timeout.
This was caused by `cilium status` taking 1.5 hours to complete. With
that step removed, let's also bump the CL2 and overall timeouts to
prevent aborting good test runs.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
bump scale-test actions to main

bump scalet-test actions to main to pick up new validate-cluster action
with 60s interval polling.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
wait for ready after creating each instance group

Nodeup began to fail on many hundreds of nodes, which required running
`kops update` to update the nodeconfig hash in the instance metadata.

Once this was done, about 1.7k nodes joined the cluster immediately and
brought the etcd cluster down. Quorum was lost and the etcd cluster
became unrecoverable.

This adds a validate-cluster wait after each step, hopefully avoiding
DDoSing the etcd cluster, and since the create-instance-group also
performs a `kops update` as the last step, the state should not become
out of sync with the config.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
update create-cluster to use 500gb etcd volumes

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

thorn3r committed:
drop FCFS slice mode

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>
